### ISSUE TITLE
Ensure alphamax linesearch is set for AcceleratedGradientDescent

### ISSUE THREAD

**Issue by anriseth** [JuliaNLSolvers/Optim.jl#458]
Accelerated Gradient Descent is very sensitive to linesearch, and I don't think we ensure that the requirements are met at the moment. 

See the description in http://stronglyconvex.com/blog/accelerated-gradient-descent.html,
and the TODO in `accelerated_gradient_descent.jl`
``` jl
# TODO: Need to specify alphamax on each iteration
# Flip notation relative to Duckworth
# Start with x_{0}
# y_{t} = x_{t - 1} - alpha g(x_{t - 1})
# If converged, return y_{t}
# x_{t} = y_{t} + (t - 1.0) / (t + 2.0) * (y_{t} - y_{t - 1})
```

**Comment by johnmyleswhite** [comment]
I think we should remove this method. I am very skeptical that it is useful.

**Comment by pkofod** [comment]
I think that AGD is often used in a ... sort of different manner than many of the other methods in Optim. I'm running some benchmarks to compare it on cutest problems just to see how well it fares.

**Comment by anriseth** [comment]
I never use AGD, as it seems to do poorly for most cases I've tried. Momentum Gradient Descent is seldom/never the best performing solver for my problems either.

It could just be because we don't give them a good initial guess on the line search, however, removing it(them) would make maintaining easier without a large cost IMO.

**Comment by pkofod** [comment]
> It could just be because we don't give them a good initial guess on the line search

I don't think it's too easy, as they are quite sensitive to the step size, but I can have a look around. Don't think they're used much though. I fixed a bug in AGD a year ago that make it almost never work, and no one had complained up until then!

**Comment by pkofod** [comment]
I will say that this is for problem sizes with `n<= 20` so it may not be the global suggestion, but with standard options, momentum and accelerated gradient descent seems like some of the worst on C2 problems in Cutest. Only simulated annealing and gradient descent seems to be worse or on par.

Again, there may be something we can do, but at present AGD and MGD with the defaults at least don't seem like a very efficient choice.

![f_err_cutest](https://user-images.githubusercontent.com/8431156/29707739-a2cc5894-8986-11e7-9960-cc8b3510e8c7.png)

**Comment by johnmyleswhite** [comment]
My impression is that people often use AGD as a technique for SGD-style model fitting when there's a varying objective function --as in training neural networks with minibatches of data. As it stands, I don't believe Optim is well-suited for optimizing a varying objective function.

**Comment by pkofod** [comment]
https://github.com/JuliaNLSolvers/LegacyOptim.jl now has functioning `AcceleratedGradientDescent` and `MomentumGradientDescent` exports. Could that be a solution? Move them there in case someone wants to use it, or see how it could be implemented in the Optim framework, but don't have them here, because we don't think they work that well in general (in this form, I know variants of them are widely used).

**Comment by anriseth** [comment]
> https://github.com/JuliaNLSolvers/LegacyOptim.jl now has functioning AcceleratedGradientDescent and MomentumGradientDescent exports. Could that be a solution? 
Yes, that's definitely better than just deleting them. They both seem to be used in more specialized cases, where people working on that application probably have better implementations suited for their needs anyway.

> (in this form, I know variants of them are widely used).
What do you mean by this?

**Comment by pkofod** [comment]
I mean that "accelerated gradient descent" and "momentum gradient descent"-style methods are widely used in the wild.
